python/test/unit/fem/test_custom_data.py

```diff
@@ -0,0 +1,287 @@
+ufcx_signature = ffcx.codegeneration.utils.numba_ufcx_kernel_signature
+
+
+# Helper intrinsic to cast void* to a typed pointer for custom_data
+@numba.extending.intrinsic
+def voidptr_to_float64_ptr(typingctx, src):
+    """Cast a void pointer (CPointer(void)) to a float64 pointer.
+
+    This function is used to access custom_data passed through the UFCx
+    tabulate_tensor interface. Since custom_data is passed as void*, this
+    intrinsic allows casting it to a typed float64 pointer for element access.
+
+    Args:
+        typingctx: The typing context.
+        src: A void pointer (CPointer(void)) to cast.
+
+    Returns:
+        sig: A Numba signature returning CPointer(float64).
+        codegen: A code generation function that performs the bitcast.
+
+    Example:
+        Inside a Numba cfunc kernel::
+
+            typed_ptr = voidptr_to_float64_ptr(custom_data)
+            scale = typed_ptr[0]  # Access first float64 value
+    """
+    # Accept CPointer(void) which shows as 'none*' in numba type system
+    if isinstance(src, numba.types.CPointer) and src.dtype == numba.types.void:
+        sig = numba.types.CPointer(numba.types.float64)(src)
+
+        def codegen(context, builder, signature, args):
+            [src] = args
+            # Cast void* to float64*
+            dst_type = context.get_value_type(numba.types.CPointer(numba.types.float64))
+            return builder.bitcast(src, dst_type)
+
+        return sig, codegen
+
+
+def tabulate_rank1_with_custom_data(dtype, xdtype):
+    """Kernel that reads a scaling factor from custom_data.
+
+    Note: custom_data must be set to a valid pointer before assembly.
+    """
+
+    @numba.cfunc(ufcx_signature(dtype, xdtype), nopython=True)
+    def tabulate(b_, w_, c_, coords_, local_index, orientation, custom_data):
+        b = numba.carray(b_, (3), dtype=dtype)
+        coordinate_dofs = numba.carray(coords_, (3, 3), dtype=xdtype)
+
+        # Cast void* to float64* and read the scale value
+        typed_ptr = voidptr_to_float64_ptr(custom_data)
+        scale = typed_ptr[0]
+
+        x0, y0 = coordinate_dofs[0, :2]
+        x1, y1 = coordinate_dofs[1, :2]
+        x2, y2 = coordinate_dofs[2, :2]
+
+        # 2x Element area Ae
+        Ae = abs((x0 - x1) * (y2 - y1) - (y0 - y1) * (x2 - x1))
+        b[:] = scale * Ae / 6.0
+
+    return tabulate
+
+
+def tabulate_rank2_with_custom_data(dtype, xdtype):
+    """Kernel that reads a scaling factor from custom_data for matrix assembly.
+
+    Note: custom_data must be set to a valid pointer before assembly.
+    """
+
+    @numba.cfunc(ufcx_signature(dtype, xdtype), nopython=True)
+    def tabulate(A_, w_, c_, coords_, entity_local_index, cell_orientation, custom_data):
+        A = numba.carray(A_, (3, 3), dtype=dtype)
+        coordinate_dofs = numba.carray(coords_, (3, 3), dtype=xdtype)
+
+        # Cast void* to float64* and read the scale value
+        typed_ptr = voidptr_to_float64_ptr(custom_data)
```

Comment:
Do we have any more complex examples? For example, an array of float64s index by the current cell. This is what @chrisrichardson mentioned before "a kernel might need to know which cell it's running on".